apiserver/facades/controller/firewaller/state.go

```diff
@@ -64,15 +66,23 @@ func (st stateShim) FirewallRule(service corefirewall.WellKnownServiceType) (*st
 	return api.Rule(service)
 }
 
-type Subnet interface {
-	ID() string
-	CIDR() string
+func (st stateShim) AllEndpointBindings() (map[string]map[string]string, error) {
```

Comment:
So looking at `stateShim` it's always a red flag to have two members with one or more being a veneer for essentially the same type.

Quick hack, still with errors:
```diff
diff --git a/apiserver/facades/controller/firewaller/firewaller.go b/apiserver/facades/controller/firewaller/firewaller.go
index c99947c379..ca854478ec 100644
--- a/apiserver/facades/controller/firewaller/firewaller.go
+++ b/apiserver/facades/controller/firewaller/firewaller.go
@@ -82,7 +82,7 @@ func NewStateFirewallerAPIV3(context facade.Context) (*FirewallerAPIV3, error) {
 		cloudspec.MakeCloudSpecCredentialContentWatcherForModel(st),
 		common.AuthFuncForTag(m.ModelTag()),
 	)
-	return NewFirewallerAPI(stateShim{st: st, State: firewall.StateShim(st, m)}, context.Resources(), context.Auth(), cloudSpecAPI)
+	return NewFirewallerAPI(stateShim{st}, context.Resources(), context.Auth(), cloudSpecAPI)
 }
 
 // NewStateFirewallerAPIV4 creates a new server-side FirewallerAPIV4 facade.
@@ -395,7 +395,7 @@ func (f *FirewallerAPIV3) getSpaceInfos() (network.SpaceInfos, error) {
 		return f.spaceInfos, nil
 	}
 
-	si, err := f.st.SpaceInfos()
+	si, err := f.st.AllSpaceInfos()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
diff --git a/apiserver/facades/controller/firewaller/state.go b/apiserver/facades/controller/firewaller/state.go
index 56cdf215b7..44048e350f 100644
--- a/apiserver/facades/controller/firewaller/state.go
+++ b/apiserver/facades/controller/firewaller/state.go
@@ -31,47 +31,35 @@ type State interface {
 
 	AllEndpointBindings() (map[string]map[string]string, error)
 
-	SpaceInfos() (network.SpaceInfos, error)
+	AllSpaceInfos() (network.SpaceInfos, error)
 }
 
 // TODO(wallyworld) - for tests, remove when remaining firewaller tests become unit tests.
 func StateShim(st *state.State, m *state.Model) stateShim {
-	return stateShim{st: st, State: firewall.StateShim(st, m)}
+	return stateShim{st, m}
 }
 
 type stateShim struct {
-	firewall.State
-	st *state.State
+	*state.State
+	*state.Model
 }
 
 func (st stateShim) ModelUUID() string {
-	return st.st.ModelUUID()
+	return st.ModelUUID()
 }
 
 func (st stateShim) GetMacaroon(entity names.Tag) (*macaroon.Macaroon, error) {
-	r := st.st.RemoteEntities()
+	r := st.RemoteEntities()
 	return r.GetMacaroon(entity)
 }
 
-func (st stateShim) FindEntity(tag names.Tag) (state.Entity, error) {
-	return st.st.FindEntity(tag)
-}
-
-func (st stateShim) WatchOpenedPorts() state.StringsWatcher {
-	return st.st.WatchOpenedPorts()
-}
-
 func (st stateShim) FirewallRule(service corefirewall.WellKnownServiceType) (*state.FirewallRule, error) {
-	api := state.NewFirewallRules(st.st)
+	api := state.NewFirewallRules(st.State)
 	return api.Rule(service)
 }
 
 func (st stateShim) AllEndpointBindings() (map[string]map[string]string, error) {
-	model, err := st.st.Model()
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
-	allEndpointBindings, err := model.AllEndpointBindings()
+	allEndpointBindings, err := st.Model.AllEndpointBindings()
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
@@ -82,7 +70,3 @@ func (st stateShim) AllEndpointBindings() (map[string]map[string]string, error)
 	}
 	return res, nil
 }
-
-func (st stateShim) SpaceInfos() (network.SpaceInfos, error) {
-	return st.st.AllSpaceInfos()
-}
```
Might be possible to pear down this to something more crisp.

Reply:
I have a cleanup card for the facade/client bits.  I can try refactoring the test suite as part of that work so we don't make this PR longer than it should be.